src/server/cluster/cluster_family.cc

```diff
@@ -0,0 +1,270 @@
+#include <string>
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "facade/dragonfly_connection.h"
+#include "facade/error.h"
+#include "server/command_registry.h"
+#include "server/conn_context.h"
+#include "server/dflycmd.h"
+#include "server/error.h"
+#include "server/replica.h"
+#include "server/server_family.h"
+#include "server/server_state.h"
+
+ABSL_FLAG(std::string, cluster_mode, "",
+          "Cluster mode supported."
+          "default: \"\"");
+ABSL_FLAG(std::string, cluster_announce_ip, "", "ip that cluster commands announce to the client");
+
+ABSL_DECLARE_FLAG(uint32_t, port);
+
+namespace dfly {
+namespace {
+
+using namespace std;
+using CI = CommandId;
+
+void BuildClusterSlotNetworkInfo(ConnectionContext* cntx, std::string_view host, uint32_t port,
+                                 std::string_view id) {
+  constexpr unsigned int kNetworkInfoSize = 3;
+
+  (*cntx)->StartArray(kNetworkInfoSize);
+  (*cntx)->SendBulkString(host);
+  (*cntx)->SendLong(port);
+  (*cntx)->SendBulkString(id);
+}
+
+}  // namespace
+
+ClusterFamily::ClusterFamily(ServerFamily* server_family) : server_family_(server_family) {
+  CHECK_NOTNULL(server_family_);
+  string cluster_mode = absl::GetFlag(FLAGS_cluster_mode);
+
+  if (cluster_mode == "emulated") {
+    is_emulated_cluster_ = true;
+  } else if (cluster_mode == "yes") {
+    cluster_config_ = std::make_unique<ClusterConfig>(server_family_->master_id());
+  } else if (!cluster_mode.empty()) {
+    LOG(ERROR) << "invalid cluster_mode. Exiting...";
+    exit(1);
+  }
+}
+
+bool ClusterFamily::IsEnabledOrEmulated() const {
+  return is_emulated_cluster_ || ClusterConfig::IsClusterEnabled();
+}
+
+string ClusterFamily::BuildClusterNodeReply(ConnectionContext* cntx) const {
+  ServerState& etl = *ServerState::tlocal();
+  auto epoch_master_time = std::time(nullptr) * 1000;
+  if (etl.is_master) {
+    std::string cluster_announce_ip = absl::GetFlag(FLAGS_cluster_announce_ip);
+    std::string preferred_endpoint =
+        cluster_announce_ip.empty() ? cntx->owner()->LocalBindAddress() : cluster_announce_ip;
+    auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
+    auto my_port = absl::GetFlag(FLAGS_port);
+    const char* connect_state = vec.empty() ? "disconnected" : "connected";
+    std::string msg = absl::StrCat(server_family_->master_id(), " ", preferred_endpoint, ":",
+                                   my_port, "@", my_port, " myself,master - 0 ", epoch_master_time,
+                                   " 1 ", connect_state, " 0-16383\r\n");
+    if (!vec.empty()) {  // info about the replica
+      const auto& info = vec[0];
+      absl::StrAppend(&msg, etl.remote_client_id_, " ", info.address, ":", info.listening_port, "@",
+                      info.listening_port, " slave 0 ", server_family_->master_id(), " 1 ",
+                      connect_state, "\r\n");
+    }
+    return msg;
+  } else {
+    auto replica = server_family_->GetReplica();
```

Comment:
The changes with MutexGuardedObject are because we use the info struct here?
If so I prefer having GetReplicaInfo function in server_family_

Reply:
Yes, it's for info and for getting the replica's master.
I am happy to expose that function in `ServerFamily`, but I do think that it's safer like this (mutex is forced)

Reply:
I would like this PR to be more simple for review. So if you think we need a change to enforce mutex lets discuss it and will decide if to create a different PR for this. But for now please just expose the ReplicaGetInfo and use the mutex as done before

Reply:
Done.